### DISCUSSION TITLE
ALE config file?

### DISCUSSION THREAD

**Discussion by spenserblack** [w0rp/ale#4099] [Ideas]
## Motivation

I have a TypeScript project, and Deno installed. I do *not* want Deno to be used as a linter for this project. But I don't want to disable Deno in my vim configuration in case I work on a Deno project.

## Idea

Like how [EditorConfig](https://github.com/editorconfig/editorconfig-vim) can read a file in a repository (root file has `root = true` to tell plugin not to look any higher in the tree), I wonder if it would be reasonable for a config file for ALE to set configurations per repository/directory instead of per buffer/pattern.

**Comment by w0rp** [comment]
This has been raised many times, and my recommendation is to use whatever Vim plugin to provide configuration specific to Vim per project, and use that to set options for ALE. ALE doesn't need to have its own implementation of project configuration files when other plugins can handle that.

**Reply by spenserblack** [reply]
Thanks for your quick response! Sorry for raising a duplicate, I guess I didn't search well enough.

If it has been raised many times, perhaps it belongs in the FAQ? Something like
Q: How can I configure ALE for specific repositories/directories?
A: Use a plugin (maybe offer suggestions)

**Reply by w0rp** [reply]
My typical recommendation is not to set linter settings for projects that aren't in a configuration file for the linter, so people who edit code in other editors get the same settings. I personally tend to just set project-specific settings in my ftplugin files by checking the current path. Others probably use project-specific files.

**Reply by pbnj** [reply]
could this be solved with a project-specific `.vimrc` file with `let g:ale_linters_ignore = {'typescript': ['deno']}`?

Edit: requires enabling `'exrc'` option in your `~/.vimrc` (see `:help 'exrc'`)

**Reply by redbmk** [reply]
For the most part I've been able to use a single config that works everywhere by not installing things globally. For example if a project uses `prettier`, ALE will find it in `node_modules` for that project and use it, but for other projects it won't run because I don't have `prettier` globally installed. This works for python as well as long as I use separate virtualenvs for each project (e.g. via poetry, pipenv, or just a `.venv` folder).

However, since `deno` is installed globally I haven't figured out how to account for this. I'd prefer not to have to go put a separate .vimrc in every project that's not using deno. It seems like there's no obvious way to tell the root of a deno project. When I run `deno init` it just creates `main.ts` and `main_test.ts` files.

I was toying with this, but `npm root` is very slow and this pauses the buffer every time you enter a TS file, so this isn't really working for me

```vimrc
    function! DetectDeno()
      let b:has_root = system('[ -d "$(npm root)" ]')
      if v:shell_error > 0
        let g:ale_linters_ignore = {}
      else
        let g:ale_linters_ignore = {'typescript': ['deno']}
      endif
    endfunction

    au BufEnter,BufNewFile,BufRead *.ts call DetectDeno()
```

The help for `'exrc'` pretty much says not to use it and recommends using an autocommand to check for certain directories.

Sounds like a couple solutions might be to use a) an autocommand to set the deno linter only for deno projects, which would involve updating that list any time you start a new deno project, b) writing some kind of significantly faster script for detecting that you're in an npm project, or c) manually turn on and off deno linting, maybe via some `nmap` binding, but that might be tough if you're switching back and forth between projects in the same vim instance.

Also none of this really addresses the fixers. It doesn't look like there's a similar `g:ale_fixers_ignore` setting. It would be nice to not have to update the full list of fixers. I guess actually this is where @w0rp's suggestion of using autocommands comes in to play, so instead of having `g:ale_fixers` look like:

```vimrc
let g:ale_fixers = {
  \ 'javascript': ['prettier', 'eslint'],
  \ 'rust': ['rustfmt'],
  " etc
\}
```

It would look something more like this?:

```vimrc
au FileType javascript let g:ale_fixers = {'javascript', ['prettier']}
au FileType rust let g:ale_fixers = {'rust', ['rustfmt']}
au FileType rust let g:ale_linters = {'rust', ['cargo', 'analyze']}
au BufEnter 
```

or I guess having separate vim/lua files in `after/ftplugin` that set those based on filetype... Not sure how to detect a folder though in case you wanted to say have `~/code/deno/**/*.ts` do something different than the rest.

**Reply by w0rp** [reply]
When you have to configure ALE for a specific project and can't commit that configuration to the project I recommend using the `b:` variables in an ftplugin file. Most of the ALE options can be set with buffer-local variables. I sometimes set some `b:` ALE variables based on the path in ftplugin files.

**Reply by spenserblack** [reply]
Deno is the problem for me, because sometimes you *do* want to use it for a TypeScript file, but sometimes you don't want to use it. I don't think that can be decided except per-project, not per-filetype.

**Reply by Flimm** [reply]
ALE doesn't just integrate linters, it also integrates fixers and formatters. Now, when I edit a file in another project, the whole file gets reformatted with different indentation or maximum line length. It would be great if I could easily configure ALE to not run fixers in some directories or projects. I can tolerate invalid linting more easily than I can tolerate having to do undo automatic fixing.

**Reply by Flimm** [reply]
I just discovered Neovim's built-in `exrc` option, which lets you automatically load a `.nvim.lua` file in the current directory, if it is marked as trusted. See my comment here: https://github.com/dense-analysis/ale/issues/832#issuecomment-2210377737

**Reply by w0rp** [reply]
Hmm. I'm thinking about this again. Could you open an issue for ALE adding support for specifically EditorConfig and tag me? I think that would be entirely reasonable, and a good developer experience.

**Reply by Flimm** [reply]
I'm not sure if `.editorconfig` files are meant to contain configuration that are specific to particular editors. It's supposed to be configuration that works across all text editors, in my opinion. I haven't given it much thought, that wasn't what OP was asking for, they were asking for support for a file like `.ale` or something, inspired by `.editorconfig`. If you're saying that vim-ale should understand the standard EditorConfig options like `indent_size`, `tab_width` and so on, that's interesting, but I don't think it is what OP was asking for.

**Reply by spenserblack** [reply]
Coming back to this discussion I honestly don't know why I brought up EditorConfig specifically as an example. Especially since an `.editorconfig` file is kind of the opposite of an editor-specific (not just that, but plugin-specific) config file. I still believe allowing for a project-wide config file is useful, but since this discussion is a duplicate I won't push the issue much further. I'm having trouble finding the duplicates, though, as my search is returning a lot of false positives.

But I believe @w0rp was talking about making ALE be able to read from `.editorconfig` (this sounds like a great idea to me BTW). If ALE starts supporting EditorConfig then you might want to see about adding ALE to EditorConfig's list of 3rd-party plugins so it's discoverable 🙂

**Reply by w0rp** [reply]
I am often in the position of skipping to the conclusion without explaining the logic. The OP is asking for a very different thing that I suggest is already better handled by writing `ftplugin` code, but I have had a brainwave from this thread. 🧠  Discovering and executing Vim or Lua scripts presents the problem of apply-eval style security issues. Very hard to implement loading a shared configuration file via a scripting language without adding security concerns. (Not that ALE is that security focused anyway, as we find and execute pretty random executables.)

1. [EditorConfig](https://editorconfig.org/) is terrificially easy to write a clean room parser for, is an already accepted standard, and keeps to a strict and simple set of rules.
1. Neovim [already supports EditorConfig](https://neovim.io/doc/user/editorconfig.html).
     1. We can potentially skip all EditorConfig logic in ALE in new enough Neovim versions and rely instead on the Vim settings Neovim will set as a result of its own EditorConfig handling.
     1. The implementation across the codebase and with specific fixes for particular linters and fixers can use the Vim settings _instead of_ pulling values _directly from_ EditorConfig so all users benefit, _even those without EditorConfig files_.
1. Automatic EditorConfig handling and modification of settings based on EditorConfig can become a general-purpose feature of ALE all users get by default bundled in with ALE.
    1. We, as we always do, will add a setting to disable this feature.
1. There's a single filename to search for (`.editorconfig`), and we can cache the probably _synchronous_ logic to find configs with a TTL where files are not found to avoided repeated IO.
1. The files are likely to also exist in projects, so ALE starts behaving better automatically in some cases.
1. Where people must be trained to add shared configs, they are not "ALE-specific," but actually teach people to use a much wider and more commonly accepted standard.
1. Ongoing maintenance won't become a complicated nightmare, which is a huge plus.

**Reply by w0rp** [reply]
https://github.com/editorconfig/editorconfig-vim Doing a little more research, this is already done! Just use this one. I'll just open an issue to make sure ALE respects line length in linters and fixers, as configured in Vim, and to document to use this plugin if you aren't already using a new enough Vim version.
